### PR TITLE
19031 edit

### PR DESCRIPTION
**Pull request by troygiorshev**
Made most of the suggested changes from #228.

### CODE REVIEW

**Comment by troygiorshev** [comment]
There are two suggested changes that I held back on including, let's discuss them here as opposed to in #228.

> Suggest removing " (this is also the first commit of PR 19031)", because in the PR description it is now the second commit.

It is the second "step" on the roadmap, but is the first commit of PR 19031.  I've included this to help avoid confusion for people who aren't familiar with the concept of a roadmap PR.

> Remove "Because of this, please review [BIP 155](https://github.com/bitcoin/bips/blob/9286b5254317d9e73fb25c5f0acd2b2d9937843e/bip-0155.mediawiki) at any of the links in these notes as they reflect the current status of the BIP."

I don't want people reviewing BIP 155 on master as it's significantly different, especially in regards to the "signaling support and compatibility" section, something that I think is worth discussing.  Without this sentence I don't think anyone will realize that I've linked a modified version of the BIP in the notes.  (Or, worse, they won't realize that a modified version of the BIP exists at all).

I've linked BIP 155 excessively both to drive this point home and to "accidentally" help anyone who doesn't read this sentence.  (That said, I agree with your suggestion to only link PR 19031 once and have made those changes).

File: _posts/2020-07-15-#19031.md [hunk 1]
```diff
@@ -12,17 +12,22 @@ commit:
...
-  v3](https://github.com/bitcoin/bitcoin/issues/18884), along with many other
-  different types of networks.
+* The `addr` message  relays connection information, including IP addresses, for peers on the Bitcoin network.  `addr` uses IPv6 addresses; IPv4 addresses are provided as IPv4-mapped IPv6 addresses.  More information [here](https://btcinformation.org/en/developer-reference#addr).
+
+* The BIP 155 `addrv2` messages would allow not only longer addresses, but also variable-length ones for future extensibility.
+
+* Bitcoin has Tor integration.  Relevant to PR 19031 is the fact that Tor v2 addresses are 16 characters long, nicely fitting in the 16 bytes needed for an IPv6 address.  Tor v3 ([meta issue here](https://github.com/bitcoin/bitcoin/issues/18884)), AKA Next Gen Onion Services, has been out since 2017.  It requires 56-character addresses and therefore needs `addrv2`.  For more information on Tor v3, see [this page](https://trac.torproject.org/projects/tor/wiki/doc/NextGenOnions) on the Tor Bug Tracker and Wiki site.
+
+* PR 19031 is large and is still a WIP.  There's no reason to review all of it in detail.  Here is a suggested approach:
```
Comment (vasild) [inline _posts/2020-07-15-#19031.md:25]: 19031 is not a WIP. It can be merged anytime if it gathers enough ACKs. If somebody has the resources - it makes perfect sense to review all of it too.

File: _posts/2020-07-15-#19031.md [hunk 2]
```diff
@@ -12,17 +12,22 @@ commit:
...
+  * Review [PR 19360](https://github.com/bitcoin/bitcoin/pull/19360) (this is also the first commit of PR 19031).
+  * Read [BIP 155](https://github.com/bitcoin/bips/blob/9286b5254317d9e73fb25c5f0acd2b2d9937843e/bip-0155.mediawiki).  There is ongoing discussion on [BIP 155](https://github.com/bitcoin/bips/blob/9286b5254317d9e73fb25c5f0acd2b2d9937843e/bip-0155.mediawiki) happening at [BIPs PR 907](https://github.com/bitcoin/bips/pull/907).  Because of this, please review [BIP 155](https://github.com/bitcoin/bips/blob/9286b5254317d9e73fb25c5f0acd2b2d9937843e/bip-0155.mediawiki) at any of the links in these notes as they reflect the current status of the BIP.
+  * See [this comment](https://github.com/bitcoin/bips/pull/907#issuecomment-611924492) in BIP PR 907 about how to signal support for `addrv2`.  The final decision can be seen [here](https://github.com/bitcoin/bips/blob/9286b5254317d9e73fb25c5f0acd2b2d9937843e/bip-0155.mediawiki#Signaling_support_and_compatibility).
```
Comment (vasild) [inline _posts/2020-07-15-#19031.md:28]: `s/The final decision/The current decision/` -- nothing is final. I gather this can change any time, even after all the code gets merged...

File: _posts/2020-07-15-#19031.md (hunk 1 above)
Comment (troygiorshev) [reply _posts/2020-07-15-#19031.md:25]: I'll reword this.  My goal is to help people who don't have a lot of time and would otherwise be intimidated by the size of 19031.

**Comment by jonatack** [comment]
Thanks @troygiorshev and @vasild! Previewing the changes now.

File: _posts/2020-07-15-#19031.md [hunk 3]
```diff
@@ -12,17 +12,22 @@ commit:
 
 ## Notes
 
-* [PR 19031](https://github.com/bitcoin/bitcoin/pull/19031) is the
-  implementation of [BIP
-  155](https://github.com/bitcoin/bips/blob/master/bip-0155.mediawiki), the
-  `addrv2` message.
+* [PR 19031](https://github.com/bitcoin/bitcoin/pull/19031) is a proposed implementation of the [BIP 155](https://github.com/bitcoin/bips/blob/9286b5254317d9e73fb25c5f0acd2b2d9937843e/bip-0155.mediawiki) `addrv2` messages, a new p2p message format proposed in early 2019 by Wladimir J. van der Laan to gossip longer node addresses. This is required to support new-generation Onion addresses, I2P, and potentially other networks that have longer endpoint addresses than fit in the 128 bits of the current addr message.
```
Comment (jonatack) [inline _posts/2020-07-15-#19031.md:15]: nit: s/addr/`addr`/ at the end of the sentence

File: _posts/2020-07-15-#19031.md [hunk 4]
```diff
@@ -12,17 +12,22 @@ commit:
...
 
-* The smaller PR that's currently being worked on, and should be reviewed
-  first, is [PR 19360](https://github.com/bitcoin/bitcoin/pull/19360).
+* The proposed implementation in PR 19031 was turned in a roadmap and separated into a series of smaller PRs to ease review. The first one in the series, [PR 19351](https://github.com/bitcoin/bitcoin/pull/19351) "test: add an edge case test for CSubNet", has been merged. [PR 19360](https://github.com/bitcoin/bitcoin/pull/19360) "net: improve encapsulation of CNetAddr" is the second PR in the series and currently under review.
```
Comment (jonatack) [inline _posts/2020-07-15-#19031.md:17]: s/turned in/converted to/

File: _posts/2020-07-15-#19031.md [hunk 5]
```diff
@@ -12,17 +12,22 @@ commit:
...
 
-* Ultimately this will allow us to work with [Tor
-  v3](https://github.com/bitcoin/bitcoin/issues/18884), along with many other
-  different types of networks.
+* The `addr` message  relays connection information, including IP addresses, for peers on the Bitcoin network.  `addr` uses IPv6 addresses; IPv4 addresses are provided as IPv4-mapped IPv6 addresses.  More information [here](https://btcinformation.org/en/developer-reference#addr).
+
+* The BIP 155 `addrv2` messages would allow not only longer addresses, but also variable-length ones for future extensibility.
+
+* Bitcoin has Tor integration.  Relevant to PR 19031 is the fact that Tor v2 addresses are 16 characters long, nicely fitting in the 16 bytes needed for an IPv6 address.  Tor v3 ([meta issue here](https://github.com/bitcoin/bitcoin/issues/18884)), AKA Next Gen Onion Services, has been out since 2017.  It requires 56-character addresses and therefore needs `addrv2`.  For more information on Tor v3, see [this page](https://trac.torproject.org/projects/tor/wiki/doc/NextGenOnions) on the Tor Bug Tracker and Wiki site.
```
Comment (jonatack) [inline _posts/2020-07-15-#19031.md:23]: s/AKA/aka/

File: _posts/2020-07-15-#19031.md [hunk 6]
```diff
@@ -12,17 +12,22 @@ commit:
...
+
+* PR 19031 a large roadmap PR.  For the purposes of this PR Review Club, here is a suggested approach:
```
Comment (jonatack) [inline _posts/2020-07-15-#19031.md:25]: s/a large/is a large/  ... though I would maybe say "is fairly large." since we've already said at least once that it is a roadmap.

Perhaps for the second sentence: "Here is a suggested approach for study and review:"

File: _posts/2020-07-15-#19031.md (hunk 4 above)
Comment (jonatack) [inline _posts/2020-07-15-#19031.md:17]: Oops, we write "proposed implementation" at the start of both first two paragraphs. Suggestion:

"To ease review, PR 19031 was turned into a roadmap and separated into a series of smaller PRs."

File: _posts/2020-07-15-#19031.md [hunk 7]
```diff
@@ -12,17 +12,22 @@ commit:
...
+  * Review [PR 19360](https://github.com/bitcoin/bitcoin/pull/19360) (this is also the first commit of PR 19031).
+  * Read [BIP 155](https://github.com/bitcoin/bips/blob/9286b5254317d9e73fb25c5f0acd2b2d9937843e/bip-0155.mediawiki).  There is ongoing discussion on [BIP 155](https://github.com/bitcoin/bips/blob/9286b5254317d9e73fb25c5f0acd2b2d9937843e/bip-0155.mediawiki) happening at [BIPs PR 907](https://github.com/bitcoin/bips/pull/907).  Because of this, please review [BIP 155](https://github.com/bitcoin/bips/blob/9286b5254317d9e73fb25c5f0acd2b2d9937843e/bip-0155.mediawiki) at any of the links in these notes as they reflect the current status of the BIP.
+  * See [this comment](https://github.com/bitcoin/bips/pull/907#issuecomment-611924492) in BIP PR 907 about how to signal support for `addrv2`.  The current decision can be seen [here](https://github.com/bitcoin/bips/blob/9286b5254317d9e73fb25c5f0acd2b2d9937843e/bip-0155.mediawiki#Signaling_support_and_compatibility).
```
Comment (jonatack) [inline _posts/2020-07-15-#19031.md:28]: s/BIP PR/BIPs PR/

Comment (jonatack) [reply _posts/2020-07-15-#19031.md:28]: (In Bitcoin Optech the same terminology is used: BIP XXX refers to a BIP, BIPs XXX refers to a PR in the BIPs repo.)

File: _posts/2020-07-15-#19031.md (hunk 4 above)
Comment (troygiorshev) [reply _posts/2020-07-15-#19031.md:17]: oof thanks

File: _posts/2020-07-15-#19031.md (hunk 5 above)
Comment (troygiorshev) [reply _posts/2020-07-15-#19031.md:23]: I universally capitalize initialisms, but when searching the bitcoin repo I see 88 results of "aka" and 0 for "AKA" or "a.k.a.", so I'll make this change.

Comment (jonatack) [reply _posts/2020-07-15-#19031.md:23]: I agree with your method, grepping the repo to decide is an excellent habit. :100:

Comment (jonatack) [reply _posts/2020-07-15-#19031.md:23]: (a.k.a. might be even better)

File: _posts/2020-07-15-#19031.md (hunk 7 above)
Comment (troygiorshev) [reply _posts/2020-07-15-#19031.md]: Finally saying it out loud "Bitcoin Improvement Proposals Pull Request" I agree that it's better 😄  Thanks!

**Comment by troygiorshev** [comment]
Thanks @jonatack @vasild for the review!  @jnewbery I think this is ready for merge.

File: _posts/2020-07-15-#19031.md [hunk 8]
```diff
@@ -12,17 +12,22 @@ commit:
 
 ## Notes
 
-* [PR 19031](https://github.com/bitcoin/bitcoin/pull/19031) is the
-  implementation of [BIP
-  155](https://github.com/bitcoin/bips/blob/master/bip-0155.mediawiki), the
-  `addrv2` message.
+* [PR 19031](https://github.com/bitcoin/bitcoin/pull/19031) is a proposed implementation of the [BIP 155](https://github.com/bitcoin/bips/blob/9286b5254317d9e73fb25c5f0acd2b2d9937843e/bip-0155.mediawiki) `addrv2` messages, a new p2p message format proposed in early 2019 by Wladimir J. van der Laan to gossip longer node addresses. This is required to support new-generation Onion addresses, I2P, and potentially other networks that have longer endpoint addresses than fit in the 128 bits of the current `addr` message.
```
Comment (jonatack) [inline _posts/2020-07-15-#19031.md:15]: One last thing (sorry): s/new/next/

**Comment by jonatack** [comment]
LGTM! I can merge this now, or let you fix the last nit.

File: _posts/2020-07-15-#19031.md (hunk 8 above)
Comment (jonatack) [reply _posts/2020-07-15-#19031.md:15]: (What *is* I2P?)

Comment (jonatack) [reply _posts/2020-07-15-#19031.md]: (If it isn't obvious to everyone but me, maybe write out what it stands for)

**Comment by troygiorshev** [comment]
new -> next

File: _posts/2020-07-15-#19031.md (hunk 8 above)
Comment (troygiorshev) [reply _posts/2020-07-15-#19031.md]: Google tells me it stands for "Invisible Internet Project" (It's one of those _smart_ initialisms).  I suspect it's one of those things that people know by the abbreviation and not the full word, let's leave it :)

**Comment by jonatack** [comment]
Thanks @troygiorshev!
